kms-message/src/kms_kmip_request.c

```diff
@@ -0,0 +1,233 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kms_kmip_reader_writer_private.h"
+#include "kms_message/kms_kmip_request.h"
+#include "kms_message_private.h"
+
+#include <inttypes.h>
+#include <stdint.h>
+
+static void
+copy_writer_buffer (kms_request_t *req, kmip_writer_t *writer) {
+   const uint8_t *buf;
+   size_t buflen;
+
+   buf = kmip_writer_get_buffer (writer, &buflen);
+   req->kmip.data = malloc (buflen);
+   memcpy (req->kmip.data, buf, buflen);
+   req->kmip.len = (uint32_t) buflen;
+}
+
+kms_request_t *
+kms_kmip_request_register_secretdata_new (void *reserved,
+                                          uint8_t *data,
+                                          uint32_t len)
+{
+   /*
+   Create a KMIP Register request with a 96 byte SecretData of this form:
+
+   <RequestMessage tag="0x420078" type="Structure">
+    <RequestHeader tag="0x420077" type="Structure">
+     <ProtocolVersion tag="0x420069" type="Structure">
+      <ProtocolVersionMajor tag="0x42006a" type="Integer" value="1"/>
+      <ProtocolVersionMinor tag="0x42006b" type="Integer" value="4"/>
+     </ProtocolVersion>
+     <BatchCount tag="0x42000d" type="Integer" value="1"/>
+    </RequestHeader>
+    <BatchItem tag="0x42000f" type="Structure">
+     <Operation tag="0x42005c" type="Enumeration" value="3"/>
+     <RequestPayload tag="0x420079" type="Structure">
+      <ObjectType tag="0x420057" type="Enumeration" value="7"/>
+      <TemplateAttribute tag="0x420091" type="Structure">
+      </TemplateAttribute>
+      <SecretData tag="0x420085" type="Structure">
+       <SecretDataType tag="0x420086" type="Enumeration" value="2"/>
```

Comment:
I was unsure whether to use `Password` or `Seed` for the `SecretDataType`. http://docs.oasis-open.org/kmip/spec/v1.4/os/kmip-spec-v1.4-os.html#_Toc490660929 I chose `Seed` arbitrarily.

Attempting to use an extension value 0x800001 resulted in an error on a PyKMIP test server. cc @markbenvenuto in case you foresee a problem with using `Seed`.

Reply:
I think `Password` might be slightly better. `Seed` is for seeding random number generates.

Extenions are something that is per-KMIP server and not something a client can arbitrarily make up